src/Compilers/Core/Portable/SourceGeneration/Nodes/TransformNode.cs

```diff
@@ -44,23 +56,41 @@ public NodeStateTable<TOutput> UpdateStateTable(DriverStateTable.Builder builder
             // - Added: perform transform and add
             // - Modified: perform transform and do element wise comparison with previous results
 
-            var newTable = previousTable.ToBuilder();
+            var newTable = previousTable.ToBuilder(builder.DriverState.TrackIncrementalSteps);
 
             foreach (var entry in sourceTable)
             {
-                if (entry.state == EntryState.Removed)
+                if (entry.State == EntryState.Removed && newTable.TryRemoveEntries())
```

Comment:
I'm curious. Could we just record the step information _inside_ the `.Try*Entries()`? Ultimately those methods have access to the steps they're writing, know if tracking is enabled, and know the outcome of what is getting written to the table?

Reply:
They don't have access to the inputs by default, so I didn't want to spend the allocations on the `ImmutableArray<(IncrementalGeneratorRunStep, int)>` of inputs when step information is disabled (though I guess I could pass a `default` instance of `ImmutableArray` in that scenario).

I'll give passing a `default` instance in the "not recording steps" scenario and see how the code looks.